docs/design/2020-01-10-shuffle-executor.md

```diff
@@ -0,0 +1,253 @@
+# Shuffle Executor
+
+- Author(s):         [pingyu](https://github.com/pingyu)  (Ping Yu)
+- Last updated:  2020-01-11
+
+## Abstract
+This article will describe the design of Shuffle Executor in TiDB.
+
+## Background
+As described in [#12966](https://github.com/pingcap/tidb/issues/12966), the performance of Window operator in TiDB has much space to improve.
+
+So we designed a Shuffle Executor, which is proposed [here](https://github.com/pingcap/tidb/pull/14238#issuecomment-569880893) by [SunRunAway](https://github.com/SunRunAway) (Feng Liyuan) , using multi-thread hash grouping, to run each Window partition in parallel.
+
+Furthmore, the Shuffle Executor is designed to be a general purpose executor, which can also be used for other scenario, such as Parallel Sort-Merge Join, in the future.
+
+
+## Rationale
+[Window operator](https://dev.mysql.com/doc/refman/8.0/en/window-functions.html) separates data into independent groups by `PARTITION BY` clause, and the independency makes parallel executing possible.
+
+The rationale of Shuffle Exector is simple: It splits and shuffles data into partitions by columns in `PARTITION BY` clause, executes Window functions on partitions in a parallel manner, and finally merges the results.
+
+![shuffle-rationale](imgs/shuffle-rationale.png)
+
+## Implementation
+
+The implementation consists of 3 parts:
+1. Physical Planning.
+2. Building Executors.
+3. Executing.
+
+### Planning
+
+We find the very position and very moment for "shuffle" in Physical Planning procedure.
+
+The very position is on top of the operator which can be "shuffled", i.e. Window operator so far.
+
+The very moment, means all of the following conditions should be met:
+
+- The corresponding parallelism switch variable is on. As to Window operator, the value of variable `tidb_window_concurreny` should be more than 1.
+- The NDV (Number of Dinstict Values) of `PARTITION BY` columns should be more than 1.
+- The parallelism should be effective enough. As to Window operator, when the child of Window operator meets the sorted property requirement (i.e. enforced sorting is not necessary), the parallelism is not effective enough. See "Benchmarks" section for detail.
+
+(_Notice that Window operator implements grouping in a "streaming" manner. So if data source is not sorted on the `PARTITION BY` columns, a Sorting would be enforced. Moreover, `SORT BY` clause in Window operator requires sorted property of data source too._)
+
+When the very position and very moment is found, we inserts a new physical plan `PhysicalShuffle`, which represents the executing plan of Shuffle Executor.
```

Comment:
Is there a `LogicalShuffle` and should we let the planner decide if we use `Shuffle`?

Reply:
By now whether using `Shuffle` or not, lies on the following conditions. The 2 & 3 should be check in physical planing, I think.
1. The corresponding parallelism switch variable is on. As to Window operator, the value of variable `tidb_window_concurreny` should be more than 1.
2. The NDV (Number of Dinstict Values) of `PARTITION BY` columns should be more than 1.
3. The parallelism should be effective enough. As to Window operator, when the child of Window operator meets the sorted property requirement (i.e. enforced sorting is not necessary), the parallelism is not effective enough. See "Benchmarks" section for detail.